enhancements/installer/vsphere-ipi-multiple-datacenter-clusters.md

```diff
@@ -0,0 +1,474 @@
+    ControlPlaneConfigs []*vsphereapis.VSphereMachineProviderSpec
+    Username            string
+    Password            string
+    Cluster             string
+    ImageURL            string
+    PreexistingFolder   bool
+}
+```
+
+##### Terraform
+
+With the added information provided in the platform spec from the `Regions`
+struct that includes the `Zones` slice we can create the appropriate tags,
+tag categories and attach them to the vCenter objects.
+
+Terraform will need to change to support cloning the control plane virtual
+machines in multiple datacenters and clusters. Below is an example how we
+can accomplish this.
+
+```golang
+variable "vsphere_control_plane_zones" {
+  type = map(object({
+    datacenter   = string
+    cluster      = string
+    network      = string
+    datastore    = string
+    resourcepool = string
+  }))
+
+data "vsphere_datacenter" "region_zones_dc" {
+  for_each = var.vsphere_control_plane_zones
+  name     = each.value["datacenter"]
+}
+
+data "vsphere_compute_cluster" "region_zones_cluster" {
+  for_each      = var.vsphere_control_plane_zones
+  name          = each.value["cluster"]
+  datacenter_id = data.vsphere_datacenter.region_zones_dc[each.key].id
+}
+data "vsphere_network" "region_zones_network" {
+  for_each      = var.vsphere_control_plane_zones
+  name          = each.value["network"]
+  datacenter_id = data.vsphere_datacenter.region_zones_dc[each.key].id
+}
+```
+
+##### Terraform variables
+
+In addition to the above terraform changes the terraform variables will also
+need to be modified. Below is an example of how we will modify the existing
+code.
+
+```golang
+type zone struct {
+    Datacenter string `json:"datacenter"`
+    Cluster string `json:"cluster"`
+    Network string `json:"network"`
+    Datastore string `json:"datastore"`
+    ResourcePool string `json:"resourcepool"`
+}
+type config struct {
+    VSphereControlPlaneZones *map[string]zone `json:"vsphere_control_plane_zones"`
+}
+
+func main() {
+    zones := &map[string]zone{
+        "master-0": {
+            Datacenter: "DC1",
+            Cluster: "C1",
+            Datastore: "DC1_DS1",
+            Network: "10-5-132-0-25",
+            ResourcePool: "/DC1/host/DC1_C1/Resources",
+        },
+    }
+    cfg := &config{VSphereControlPlaneZones:zones}
+    jsonBytes, _ := json.MarshalIndent(cfg, "", "  ")
+```
+
+##### Output from above, for example
+
+```golang
+$ cat zones.json| jq
+{
+  "vsphere_control_plane_zones": {
+    "master-0": {
+      "datacenter": "DC1",
+      "cluster": "C1",
+      "network": "10-5-132-0-25",
+      "datastore": "DC1_DS1",
+      "resourcepool": "/DC1/host/DC1_C1/Resources"
+    }
+  }
+}
+```
+
+#### Machine and MachineSet
+
+The control plane
+[Machines](https://github.com/openshift/installer/blob/b0b96468893db2240e82ba2aa0935679c8c49201/pkg/asset/machines/vsphere/machines.go#L19-L64)
+will need to be modified to create a Machine per `region` and `zone` pair.
```

Comment:
The zonal information is picked up by the CCM via tags and applied to the node objects do you see a reason to extend the machine api for that?

Reply:
Today you can configure the zonal information on each Machine by setting the workspace, we could also offer users the option to request a Machine in zone foo and have the Machine API read this from something like the Infrastructure object. This is a pattern that vSphere uses in Cluster API upstream for example. Possibly not required for the initial integration but worth considering so that we don't prevent this in the future if we decide we do want to do it

Reply:
There is no zonal information in the workspace currently unless I am not looking in the correct location.

https://github.com/openshift/api/blob/864d80f741744d2dc831590de8b2fc0b32f8779d/machine/v1beta1/types_vsphereprovider.go#L99-L115

That seems to be me to be out of scope for this initial effort.